DataScience/Analytics/beta/check_performance.py

```diff
@@ -0,0 +1,266 @@
+    msk = deduped['matches'] == True
+    correct = deduped.loc[msk]
+    number_of_correct = len(correct.index)
+
+    print('Top Ranking Match is Correct:', number_of_correct)
+    results.append(number_of_correct)
+
+    # find those ids where the highest scored match is not the correct match and UPRNs were found
+    top_id_is_not_correct = deduped.loc[~msk & deduped['UPRN_beta'].notnull()]['ID_original']
+
+    correct_in_set = []
+    incorrect = []
+    if len(top_id_is_not_correct.values) > 0:
+        for not_correct_id in top_id_is_not_correct.values:
+            # check if the correct answer is within the set
+            values = data.loc[data['ID_original'] == not_correct_id]
+            sum_of_matches = np.sum(values['matches'].values)
+
+            if sum_of_matches >= 1:
+                correct_in_set.append(not_correct_id)
+            else:
+                incorrect.append(not_correct_id)
+
+        print('Correct Match in the Set of Returned Addresses:', len(correct_in_set))
+        print('Correct Match not in the Set:', len(incorrect))
+
+    results.append(len(correct_in_set))
+    results.append(len(incorrect))
+
+    if verbose:
+        print(correct_in_set)
+        print(incorrect)
+
+    return results
+
+
+def _generate_performance_figure(all_results, filename, width=0.35):
+    """
+    Generate a simple bar chart to show the results.
+
+    :param all_results: a list containing all the results as computed by the _check_performance method
+    :type all_results: list
+    :param filename: name of the output file
+    :type filename: str
+    :param width: fractional width of the bars
+    :type width: float
+
+    :return: None
+    """
+    all_results_names = ['Input Addresses', 'Not Matched', 'Top Ranking Match', 'Within the Set', 'Incorrect']
+    location = np.arange(len(all_results))
+
+    fig = plt.figure(figsize=(12, 10))
+    plt.title('Beta Linking Service ({})'.format(datetime.datetime.now().strftime("%Y-%m-%d %H%M%S")))
+    ax = fig.add_subplot(1, 1, 1)
+
+    max_bar_length = max(all_results)
+    plt.barh(location, all_results, width, color='g', alpha=0.6)
+
+    for patch in ax.patches:
+        if patch.get_x() < 0:
+            continue
+
+        n_addresses = int(patch.get_width())
+        ratio = n_addresses / max_bar_length
+
+        if ratio > 0.3:
+            ax.annotate("%i" % n_addresses, (patch.get_x() + patch.get_width(), patch.get_y()),
+                        xytext=(-95, 18), textcoords='offset points', color='white', fontsize=24)
+        else:
+            ax.annotate("%i" % n_addresses, (patch.get_x() + patch.get_width(), patch.get_y()),
+                        xytext=(10, 18), textcoords='offset points', color='black', fontsize=24)
+
+    plt.xlabel('Number of Addresses')
+    plt.yticks(location, all_results_names)
+    plt.xlim(0, ax.get_xlim()[1] * 1.02)
+    plt.tight_layout()
+    plt.savefig(filename)
+    plt.close()
+
+
+def main(path):
```

Comment:
I generally prefer writing top-level function at the top and then writing other functions one after another in the order they appear in the code. This way it's easier to read code files as you are reading them as a news paper instead of reading it as a dictionary. (more on it: "Clean code" by Robert C. Martin)